feat: web share.

creampnx-x opened the pull request:
## Web Share

### Support

+ firefox
    + mac os: can be enable, default false.
    + windows: can be enable, default false.
    + android: true, but not `files`.
+ chrome, edge
    + mac os: `false`.
    + windows: true, but not `files`.
    + android: true.
+ safari
    + mac os, ios: true.

details: 
+ https://caniuse.com/?search=navigator.canShare
+ https://caniuse.com/mdn-api_navigator_canshare_data_files_parameter

### createWebShare

#### screen shot
Test on Mac os Safari. 
1. Initial Appearance
<img width="343" alt="截屏2022-12-01 15 59 01" src="https://user-images.githubusercontent.com/73047992/204997658-056a1ee8-4bbb-4627-8cb2-9726f78f1eeb.png">

2. click share button.
<img width="1309" alt="截屏2022-12-01 15 53 17" src="https://user-images.githubusercontent.com/73047992/204997042-8a935c74-9dd3-4299-88d3-2bc26b31a80f.png">

3. after sharing, the info change: `successful sharing.`
<img width="499" alt="截屏2022-12-01 15 54 04" src="https://user-images.githubusercontent.com/73047992/204997390-20a4343c-c0b9-4807-af39-7042353ea2c2.png">

davedbase commented:
This is a wonderful contribution. Thank you so much! I have a couple suggestions:

- [ ] I'd prefer not to put this in the `browser` package. I recommend consolidating it in the `share` package that exists so that createWebShare lives beside createSocialShare. Another option is a new package called `webshare`
- [ ] I would simplify the demo and use the signal from the primitive directly instead of a createEffect that updates other signals.
- [ ] Would it make sense to have a makeWebShare and compose it with createWebShare? or is this API simple enough that producing either types of sharing primitives wont be necessary?

creampnx-x commented:
Thank you very much for your review. I'm going to try this done this weekend.

creampnx-x commented:
## Change
+ I added it to the `web-share` package, because the `share` package is exported by default.
+ I refactored the `createWebShare` using `createEffect` and `from`, so we can use the store of primitives directly. Besides, the return of primitives is not a signal but a store which has `status` and `failing message` properties.
+ I added `makeWebShare` which hasn't reactivity and compose it with `createWebShare`.

## Screen shots
<img width="400" alt="截屏2022-12-03 21 43 35" src="https://user-images.githubusercontent.com/73047992/205443910-8eade6db-50a0-44ec-99b4-c1e00d43218b.png"><img width="400" alt="截屏2022-12-03 21 44 02" src="https://user-images.githubusercontent.com/73047992/205443930-7ebbeef0-e868-40a6-bde5-5035080ccc10.png">

atk commented on packages/web-share/src/index.ts:
```diff
@@ -0,0 +1,87 @@
+              setStatus({ status: false, message: e.toString() });
+            });
+        },
+        { defer: deferInitial || true }
```
`deferInitial` will make no difference here. What you want is `deferInitial !== false`.

creampnx-x replied:
Done.

thetarnav reviewed (commented):
Thank you, this is a very nice addition.
Here are a couple of suggestions from looking through the code.

creampnx-x commented on packages/web-share/src/index.ts:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  const share = makeWebShare();
+  if (data)
```
I have changed it to just be required, because web share can not be actively paused.
